specs/stein/approved/amd-sev-libvirt-support.rst

```diff
@@ -0,0 +1,367 @@
+data from another VM running on the same hardware without being
+detected.  What was once a theoretical attack has now been realized
+with the result that customers are concerned that their data may not
+be adequately protected in a cloud environment.  AMD's SEV offers a
+potential solution for this which transparently encrypts the memory of
+each VM with a unique key.  It can also calculate a signature of the
+memory contents, which can be sent to the guest owner as an
+attestation that the memory was encrypted correctly by the firmware.
+Therefore the solution can also provide some degree of protection
+against a malicious hypervisor or compromised cloud operator, although
+it should probably not be relied upon as the sole line of defense.
+
+Use Cases
+---------
+
+#. Booting an instance with a flavor containing an SEV-specific
+   `required trait extra spec
+   <https://docs.openstack.org/nova/latest/user/flavors.html#extra-specs-required-traits>`_
+   will cause that instance to be defined and booted on a compute host
+   which provides that trait.  Additionally the instance should be
+   created with additional XML in the libvirt domain definition which
+   causes QEMU to boot the instance with SEV enabled.
+
+Proposed change
+===============
+
+To achieve the simplest prototype implementation which can possibly
+work, it is proposed that initially during development and testing,
+the placement API would be used to:
+
+- `create a custom trait
+  <https://docs.openstack.org/osc-placement/latest/cli/index.html#trait-create>`_
+  specifically for the purpose of marking flavors as SEV-capable, and
+  then
+
+- `manually assign that trait
+  <https://docs.openstack.org/osc-placement/latest/cli/index.html#resource-provider-trait-set>`_
+  to each compute node which is SEV-capable.
+
+This trait would also indicate that the libvirt driver should include
+extra XML in the guest's domain definition, to enable SEV security,
+ensure that the VM uses machine type ``q35`` and UEFI, and that
+``iommu_platform=on`` is used for all virtio devices, e.g.::
+
+    <domain type='kvm'>
+      <os>
+        <type arch='x86_64' machine='pc-q35-2.11'>hvm</type>
+        <loader readonly='yes' type='pflash'>/usr/share/qemu/ovmf-x86_64-ms-4m-code.bin</loader>
+        <nvram>/var/lib/libvirt/qemu/nvram/sles15-sev-guest_VARS.fd</nvram>
+        <boot dev='hd'/>
+      </os>
+      <launchSecurity type='sev'>
+        <cbitpos>47</cbitpos>
+        <reducedPhysBits>1</reducedPhysBits>
+        <policy>0x0001</policy>
+      </launchSecurity>
+      ...
+    </domain>
+
+For reference, `the AMDSEV GitHub repository
+<https://github.com/AMDESE/AMDSEV/>`_ provides `a complete example of
+SEV-enabled domain XML
+<https://github.com/AMDESE/AMDSEV/blob/master/xmls/sample.xml>`_.
+
+This additional functionality could be implemented within
+``nova.virt.libvirt.LibvirtDriver._set_features()``, e.g. using a new
+``nova.virt.libvirt.LibvirtConfigGuestFeatureSEV`` class.
+
```

Comment:
As SEV requires a libvirt >= x, the necessary functions can be implemented with the help of the usual has_min_version decorator to be non-intrusive.

Reply:
I don't know much about that decorator; how does it work?  Would it support a cloud with a mix of SEV and non-SEV compute hosts?  Does this bring any extra benefit over just using libvirt's virConnectGetDomainCapabilities() API call as described in this spec?  I think I'd prefer to trust libvirt's own opinion of whether it can support SEV, rather than reimplementing that test.